### PR TITLE
std::future preview

### PR DESCRIPTION
**Pull request by bluejekyll**
I was hoping this could be done without pulling a dependency on async/await, but no such luck. There are some complex areas that make this painful.

### CODE REVIEW

**Comment by gfreezy** [comment]
What's the progress of this PR?  I want to do some help.

**Comment by bluejekyll** [comment]
FYI: Current progress. the base proto, client, and resolver crates are all fully passing tests with default features.

Now working on server, *tls, and the integration test crates. No interfaces have been updated yet to be async/await friendly, but getting there.

**Comment by baloo** [comment]
There is a weird file named "1" added by 6602c687b4345ef6b23c6c64515de22939068fe2, looks like an error.

**Comment by bluejekyll** [comment]
Oh, huh. I wonder how that happened. Thanks for the heads up.

**Comment by bluejekyll** [comment]
FYI, for an update on the progress of this. All crates are compiling with default features, and integration tests are all passing, except for one. There is a test that was reliant on task local storage for tracking the depth of CNAME chain lookups in the resolver. I'm working to rewrite that logic such that we can still make sure we don't follow CNAME chains beyond a reasonable level.

**Comment by bluejekyll** [comment]
Ok, all default features are now compiling, all tests passing. Max Query depth for CNAME chains is enforced again. This is getting close for a review.

**Comment by bluejekyll** [comment]
FYI, just push a rebase onto master. So any forks should be aware. I'm going to go through and review all of this change, I there are quite a few FIXMEs I sprinkled throughout, need to clean that up.

File: crates/proto/src/xfer/dns_multiplexer.rs [hunk 1]
```diff
@@ -215,20 +213,18 @@ where
...
             if !self.active_requests.contains_key(&id) {
-                return Async::Ready(id);
+                return Poll::Ready(id);
             }
         }
 
-        warn!("could not get next random query id, delaying");
-        task::current().notify();
-        Async::NotReady
+        Poll::Pending
```
Comment (baloo) [inline crates/proto/src/xfer/dns_multiplexer.rs]: isn't this gonna bubble up the pending, and never going to be called back again? (scheduler is not going to know when this is to become ready again)

Comment (bluejekyll) [reply crates/proto/src/xfer/dns_multiplexer.rs]: oh, thanks for catching this. I need to pass through the Context or a Waker to immediately wake on this.

Comment (bluejekyll) [reply crates/proto/src/xfer/dns_multiplexer.rs]: ok, this should be covered now.

**Comment by bluejekyll** [comment]
Ok. I believe this is ready to go. I’ll be merging in then will publish the first preview! We’ll want to adjust some of the APIss I’m sure.

**Comment by zonyitoo** [comment]
Still cannot pass travis' tests.

**Comment by baloo** [comment]
@bluejekyll congrats !

**Comment by bluejekyll** [comment]
Thanks, I'm just happy all the Futures work stabilized, so there shouldn't be a need to do a massive upgrade like this again. That was a slog.

**Pull request merged**